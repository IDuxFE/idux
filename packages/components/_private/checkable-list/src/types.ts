/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
