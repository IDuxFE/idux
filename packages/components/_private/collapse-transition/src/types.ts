/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
