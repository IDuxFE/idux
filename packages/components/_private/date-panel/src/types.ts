/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
