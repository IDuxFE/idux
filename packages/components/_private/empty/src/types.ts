/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
