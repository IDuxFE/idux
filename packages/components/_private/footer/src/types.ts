/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
