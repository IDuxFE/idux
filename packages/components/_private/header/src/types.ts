/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { HeaderProps as IxHeaderProps } from '@idux/components/header'
import type { VNode } from 'vue'

export interface HeaderProps {
  closable?: boolean
  closeIcon?: string | VNode
  header?: string | IxHeaderProps
  size?: 'lg' | 'md' | 'sm'
  onClose?: (evt: Event) => void
}
