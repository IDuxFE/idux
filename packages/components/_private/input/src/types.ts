/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ValidateStatus } from '@idux/cdk/forms'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { DefineComponent, InputHTMLAttributes, PropType } from 'vue'

export const inputProps = {
  addonAfter: String,
  addonBefore: String,
  borderless: {
    type: Boolean,
    default: undefined,
  },
  clearable: {
    type: Boolean,
    default: undefined,
  },
  clearIcon: String,
  clearVisible: {
    type: Boolean,
    default: undefined,
  },
  disabled: {
    type: Boolean,
    default: undefined,
  },
  focused: {
    type: Boolean,
    default: undefined,
  },
  prefix: String,
  size: String as PropType<FormSize>,
  status: String as PropType<ValidateStatus>,
  suffix: String,
  onClear: Function as PropType<(evt: MouseEvent) => void>,
} as const

export type InputProps = ExtractInnerPropTypes<typeof inputProps>
export type InputPublicProps = ExtractPublicPropTypes<typeof inputProps>
export interface InputBindings {
  getInputElement: () => HTMLInputElement | undefined
}
export type InputComponent = DefineComponent<
  Omit<InputHTMLAttributes, keyof InputPublicProps> & InputPublicProps,
  InputBindings
>
export type InputInstance = InstanceType<DefineComponent<InputProps, InputBindings>>
