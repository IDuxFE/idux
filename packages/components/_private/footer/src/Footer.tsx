/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { FooterButtonProps } from './types'
import type { VNodeChild } from 'vue'

import { computed, defineComponent, isVNode } from 'vue'

import { isFunction } from 'lodash-es'

import { IxButton } from '@idux/components/button'

import { footerProps } from './types'

export default defineComponent({
  name: 'ÉµFooter',
  props: footerProps,
  setup(props, { slots }) {
    const cancelButtonProps = computed<FooterButtonProps | undefined>(() => {
      const { cancel, cancelButton, cancelVisible, cancelLoading, cancelText } = props
      if (!cancelVisible || (!cancel && !cancelText && !cancelButton)) {
        return undefined
      }
      return {
        key: '__IDUX_FOOTER_BUTTON_CANCEL',
        text: cancelText,
        onClick: cancel,
        loading: cancelLoading,
        ...cancelButton,
      } as FooterButtonProps
    })

    const okButtonProps = computed<FooterButtonProps | undefined>(() => {
      const { ok, okButton, okLoading, okText } = props
      if (!ok && !okText && !okButton) {
        return undefined
      }
      return {
        key: '__IDUX_FOOTER_BUTTON_OK',
        text: okText,
        onClick: ok,
        loading: okLoading,
        mode: cancelButtonProps.value ? 'primary' : 'default',
        ...okButton,
      } as FooterButtonProps
    })

    return () => {
      const footerProp = props.footer
      const footerSlot = slots.footer
      if (!footerProp && !footerSlot) {
        return undefined
      }

      let children: VNodeChild
      if (footerSlot) {
        children = footerSlot(props)
      } else if (isFunction(footerProp)) {
        children = footerProp()
      } else if (isVNode(footerProp)) {
        children = footerProp
      } else {
        let buttonProps = footerProp
        if (!Array.isArray(buttonProps)) {
          const cancelButton = cancelButtonProps.value
          const okButton = okButtonProps.value
          buttonProps = []
          okButton && buttonProps.push(okButton)
          cancelButton && buttonProps.push(cancelButton)
        }
        children = buttonProps.map(item => {
          const { text, disabled, ...rest } = item
          const _text = isFunction(text) ? text() : text
          return (
            <IxButton disabled={disabled || props.disabled} {...rest}>
              {_text}
            </IxButton>
          )
        })
      }

      return <div>{children}</div>
    }
  },
})
