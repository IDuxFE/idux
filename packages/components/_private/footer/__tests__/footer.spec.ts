import { MountingOptions, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import Footer from '../src/Footer'
import { FooterProps } from '../src/types'

describe('Footer', () => {
  const FooterMount = (options?: MountingOptions<Partial<FooterProps>>) => {
    const { props, ...rest } = options || {}
    return mount(Footer, { props: { footer: true, ...props }, ...rest })
  }

  renderWork<FooterProps>(Footer, {
    props: { footer: true, okText: 'ok', cancelText: 'cancel' },
  })

  test('cancel work', async () => {
    const cancel = vi.fn()
    const wrapper = FooterMount({ props: { cancel } })

    expect(wrapper.findAll('.ix-button').length).toBe(1)

    await wrapper.find('.ix-button').trigger('click')

    expect(cancel).toBeCalled()
  })

  test('cancelButton work', async () => {
    const cancel = vi.fn()
    const wrapper = FooterMount({ props: { cancel, cancelButton: { mode: 'primary' } } })

    expect(wrapper.find('.ix-button').classes()).toContain('ix-button-primary')

    await wrapper.setProps({ cancelButton: { mode: 'dashed' } })

    expect(wrapper.find('.ix-button').classes()).not.toContain('ix-button-primary')
    expect(wrapper.find('.ix-button').classes()).toContain('ix-button-dashed')
  })

  test('cancelLoading work', async () => {
    const cancel = vi.fn()
    const wrapper = FooterMount({ props: { cancel, cancelLoading: true } })

    expect(wrapper.find('.ix-button').classes()).toContain('ix-button-loading')

    await wrapper.setProps({ cancelLoading: false })

    expect(wrapper.find('.ix-button').classes()).not.toContain('ix-button-loading')
  })

  test('cancelText work', async () => {
    let cancelText = 'cancel'
    const wrapper = FooterMount({ props: { cancelText } })

    expect(wrapper.find('.ix-button').text()).toBe(cancelText)

    cancelText = 'cancel 2'
    await wrapper.setProps({ cancelText })

    expect(wrapper.find('.ix-button').text()).toBe(cancelText)
  })

  test('cancelVisible work', async () => {
    const cancelText = 'cancel'
    const wrapper = FooterMount({ props: { cancelText, cancelVisible: false } })

    expect(wrapper.findAll('.ix-button').length).toBe(0)

    await wrapper.setProps({ cancelVisible: true })

    expect(wrapper.findAll('.ix-button').length).toBe(1)
  })

  test('footer with boolean work', async () => {
    const wrapper = FooterMount({ props: { footer: false } })

    expect(wrapper.isVisible()).toBe(false)

    await wrapper.setProps({ footer: true })

    expect(wrapper.isVisible()).toBe(true)
  })

  test('footer with buttonProps work', async () => {
    const wrapper = FooterMount({ props: { footer: [{ text: 'cancel' }, { text: 'ok' }] } })

    expect(wrapper.findAll('.ix-button').length).toBe(2)

    await wrapper.setProps({ footer: [{ text: 'ok' }] })

    expect(wrapper.findAll('.ix-button').length).toBe(1)
  })

  test('footer with slot work', async () => {
    const wrapper = FooterMount({
      props: { footer: [{ text: 'cancel' }, { text: 'ok' }] },
      slots: { footer: () => [h('div', { class: 'test-footer' }, 'cancel'), h('div', { class: 'test-footer' }, 'ok')] },
    })

    expect(wrapper.findAll('.ix-button').length).toBe(0)
    expect(wrapper.findAll('.test-footer').length).toBe(2)
  })

  test('ok work', async () => {
    const ok = vi.fn()
    const wrapper = FooterMount({ props: { ok } })

    expect(wrapper.findAll('.ix-button').length).toBe(1)

    await wrapper.find('.ix-button').trigger('click')

    expect(ok).toBeCalled()
  })

  test('okButton work', async () => {
    const ok = vi.fn()
    const wrapper = FooterMount({ props: { ok, okButton: { mode: 'dashed' } } })

    expect(wrapper.find('.ix-button').classes()).toContain('ix-button-dashed')

    await wrapper.setProps({ okButton: undefined })

    expect(wrapper.find('.ix-button').classes()).not.toContain('ix-button-dashed')
  })

  test('okLoading work', async () => {
    const ok = vi.fn()
    const wrapper = FooterMount({ props: { ok, okLoading: true } })

    expect(wrapper.find('.ix-button').classes()).toContain('ix-button-loading')

    await wrapper.setProps({ okLoading: false })

    expect(wrapper.find('.ix-button').classes()).not.toContain('ix-button-loading')
  })

  test('okText work', async () => {
    let okText = 'cancel'
    const wrapper = FooterMount({ props: { okText } })

    expect(wrapper.find('.ix-button').text()).toBe(okText)

    okText = 'cancel 2'
    await wrapper.setProps({ okText })

    expect(wrapper.find('.ix-button').text()).toBe(okText)
  })
})
