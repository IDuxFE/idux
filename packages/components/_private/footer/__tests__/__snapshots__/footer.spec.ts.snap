// Vitest Snapshot v1

exports[`Footer > render work 1`] = `
"<div><button class=\\"ix-button ix-button-primary ix-button-md\\" type=\\"button\\"><span>ok</span>
    <div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div>
  </button><button class=\\"ix-button ix-button-default ix-button-md\\" type=\\"button\\"><span>cancel</span>
    <div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div>
  </button></div>"
`;
